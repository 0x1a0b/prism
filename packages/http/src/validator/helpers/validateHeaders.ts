import { IValidation } from '@stoplight/prism-core';
import { ValidationSeverity } from '@stoplight/prism-core/types';
import { IHttpContent, IHttpHeaderParam } from '@stoplight/types/http';
import * as Ajv from 'ajv';
import { convertAjvErrors } from './convertAjvErrors';

const ajv = new Ajv({ allErrors: true, messages: true });

export function validateHeaders(
  headers: { [name: string]: string } = {},
  specs: IHttpHeaderParam[] = [],
  mediaType?: string
): IValidation[] {
  return specs.reduce<IValidation[]>((results, spec) => {
    if (!headers.hasOwnProperty(spec.name) && spec.required === true) {
      results.push({
        path: ['header', spec.name],
        name: 'required',
        summary: '',
        message: `Missing ${spec.name} header param`,
        severity: ValidationSeverity.ERROR,
      });

      // stop further checks
      return results;
    }

    const content = resolveContent(spec.content, mediaType);

    if (content && content.schema) {
      if (spec.explode === true) {
        Array.prototype.push.apply(
          results,
          explodeAndValidateAgainstSchema(headers, content.schema)
        );
      } else {
        Array.prototype.push.apply(
          results,
          validateAgainstSchema(headers[spec.name], content.schema)
        );
      }
    }

    if (spec.deprecated === true) {
      results.push({
        path: ['header', spec.name],
        name: 'deprecated',
        summary: '',
        message: `Header param ${spec.name} is deprecated`,
        severity: ValidationSeverity.WARN,
      });
    }

    return results;
  }, []);
}

function explodeAndValidateAgainstSchema(headers: { [name: string]: string }, schema: any) {
  if (schema.type !== 'object') {
    throw new Error('Explode can be only applied to header of object type');
  }

  const schemaProperties = schema.properties || {};

  const validationSubject = Object.keys(schemaProperties).reduce(
    (subject, key) => ({ ...subject, [key]: headers[key] }),
    {}
  );

  const validate = ajv.compile(schema);
  if (!validate(validationSubject)) {
    return convertAjvErrors(validate.errors, 'header', ValidationSeverity.ERROR);
  }

  return [];
}

function validateAgainstSchema(value: string, schema: any): IValidation[] {
  const validate = ajv.compile(schema);

  if (!validate(value)) {
    return convertAjvErrors(validate.errors, 'header', ValidationSeverity.ERROR);
  }

  return [];
}

function resolveContent(content: { [mediaType: string]: IHttpContent }, mediaType?: string) {
  if (!mediaType) {
    if (content.hasOwnProperty('*')) {
      return content['*'];
    }

    return null;
  }

  if (content.hasOwnProperty(mediaType)) {
    return content[mediaType];
  }

  return content['*'];
}
