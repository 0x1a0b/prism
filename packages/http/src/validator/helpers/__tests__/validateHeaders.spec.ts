import { validateHeaders } from '../validateHeaders';

describe('validateHeaders()', () => {
  describe('headers are missing', () => {
    it('returns empty validation error list', () => {
      expect(validateHeaders(undefined, [], 'application/json')).toEqual([]);
    });
  });

  describe('spec for headers is missing', () => {
    it('returns empty validation error list', () => {
      expect(validateHeaders({}, undefined, 'application/json')).toEqual([]);
    });
  });

  describe('spec is present', () => {
    describe('header is not present', () => {
      describe('spec defines it as required', () => {
        it('returns validation error', () => {
          expect(
            validateHeaders({}, [{ name: 'aHeader', required: true, content: {} }])
          ).toMatchSnapshot();
        });
      });
    });

    describe('header is present', () => {
      describe('schema is present', () => {
        describe('header is valid', () => {
          it('validates positively against schema', () => {
            expect(
              validateHeaders({ 'x-test-header': 'abc' }, [
                {
                  name: 'x-test-header',
                  content: { '*': { mediaType: '*', schema: { type: 'string' } } },
                },
              ])
            ).toEqual([]);
          });
        });

        describe('header is not valid', () => {
          it('not validates against schema', () => {
            expect(
              validateHeaders({ 'x-test-header': 'abc' }, [
                {
                  name: 'x-test-header',
                  content: { '*': { mediaType: '*', schema: { type: 'number' } } },
                },
              ])
            ).toMatchSnapshot();
          });
        });
      });

      describe('mediaType is provided', () => {
        describe('specific mediaType found in spec content list', () => {
          it('validates against more specific spec', () => {
            expect(
              validateHeaders(
                { 'x-test-header': 'abc' },
                [
                  {
                    name: 'x-test-header',
                    content: {
                      '*': { mediaType: '*', schema: { type: 'string' } },
                      'application/testson': {
                        mediaType: 'application/testson',
                        schema: { type: 'number' },
                      },
                    },
                  },
                ],
                'application/testson'
              )
            ).toMatchSnapshot();
          });
        });

        describe('specific mediaType not found in spec content list', () => {
          it('validates against default spec', () => {
            expect(
              validateHeaders(
                { 'x-test-header': 'abc' },
                [
                  {
                    name: 'x-test-header',
                    content: {
                      '*': { mediaType: '*', schema: { type: 'string' } },
                      'application/testson': {
                        mediaType: 'application/testson',
                        schema: { type: 'number' },
                      },
                    },
                  },
                ],
                'application/failson'
              )
            ).toMatchSnapshot();
          });
        });
      });

      describe('deprecated flag is set', () => {
        it('returns deprecation warning', () => {
          expect(
            validateHeaders({ 'x-test-header': 'abc' }, [
              {
                name: 'x-test-header',
                deprecated: true,
                content: {},
              },
            ])
          ).toMatchSnapshot();
        });
      });

      describe('explode flag is set', () => {
        describe('schema type is incorrect', () => {
          it('throws error', () => {
            expect(() =>
              validateHeaders({ 'x-test-header': 'abc' }, [
                {
                  name: 'x-test-header',
                  explode: true,
                  content: {
                    '*': { mediaType: '*', schema: { type: 'string' } },
                  },
                },
              ])
            ).toThrowErrorMatchingSnapshot();
          });
        });

        describe('header is valid', () => {
          it('validates positively', () => {
            expect(
              validateHeaders(
                {
                  'x-test-header-1': 'abc',
                  'x-test-header-2': 'def',
                },
                [
                  {
                    name: 'x-test-header',
                    explode: true,
                    content: {
                      '*': {
                        mediaType: '*',
                        schema: {
                          type: 'object',
                          properties: {
                            'x-test-header-1': { type: 'string' },
                            'x-test-header-2': { type: 'string' },
                          },
                        },
                      },
                    },
                  },
                ]
              )
            ).toEqual([]);
          });
        });

        describe('header is invalid', () => {
          it('not validates', () => {
            expect(
              validateHeaders(
                {
                  'x-test-header-1': 'abc',
                  'x-test-header-2': 'def',
                },
                [
                  {
                    name: 'x-test-header',
                    explode: true,
                    content: {
                      '*': {
                        mediaType: '*',
                        schema: {
                          type: 'object',
                          properties: {
                            'x-test-header-1': { type: 'string' },
                            'x-test-header-2': { type: 'number' },
                          },
                        },
                      },
                    },
                  },
                ]
              )
            ).toMatchSnapshot();
          });
        });
      });
    });
  });
});
