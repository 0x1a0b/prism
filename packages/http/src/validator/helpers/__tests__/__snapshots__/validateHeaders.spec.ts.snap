// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateHeaders() spec is present header is not present spec defines it as required returns validation error 1`] = `
Array [
  Object {
    "message": "Missing aHeader header param",
    "name": "required",
    "path": Array [
      "header",
      "aHeader",
    ],
    "severity": "error",
    "summary": "",
  },
]
`;

exports[`validateHeaders() spec is present header is present deprecated flag is set returns deprecation warning 1`] = `
Array [
  Object {
    "message": "Header param x-test-header is deprecated",
    "name": "deprecated",
    "path": Array [
      "header",
      "x-test-header",
    ],
    "severity": "warn",
    "summary": "",
  },
]
`;

exports[`validateHeaders() spec is present header is present explode flag is set header is invalid not validates 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "header",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;

exports[`validateHeaders() spec is present header is present explode flag is set schema type is incorrect throws error 1`] = `"Explode can be only applied to header of object type"`;

exports[`validateHeaders() spec is present header is present mediaType is provided specific mediaType found in spec content list validates against more specific spec 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "header",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;

exports[`validateHeaders() spec is present header is present mediaType is provided specific mediaType not found in spec content list validates against default spec 1`] = `Array []`;

exports[`validateHeaders() spec is present header is present schema is present header is not valid not validates against schema 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "header",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;
